File: quicksort.java
class QuickSort {
  public static int divide(int []num, int left, int right) {
    int base = num[left];
    while (left < right) {
      while (left < right && num[right] > base) {
        right--;
      }
      if (left < right) {
        num[left++] = num[right];
      }
      while (left < right && num[left] < base) {
        left++;
      }
      if (left < right) {
        num[right--] = num[left];
      }
    }
    num[left] = base;
    return left;
  }
  public static void sort(int []num, int left, int right) {
    if (left > right) {
      return ;
    } else {
      int base = divide(num, left, right);
      sort(num, left, base - 1);
      sort(num, base + 1, right);
    }
  }
  public static void main(String[] args) {
    int num[] = {3, 5, 6, 1, 8, 2, 9};
    sort(num, 0, 6);
    for (int item : num) {
      System.out.println(item);
    }
  }
}

File: quick-sort.js
var arr = [2, 3, 4, 1, 5, 6, 2, 4, 9];
function quickSort(left = 0, right = arr.length - 1) {
  const base = arr[left];
  while (left < right) {
    while (left < right && arr[right] > base) {
      right--;
    }
    if (left < right) {
      arr[left++] = arr[right];
    }
    while (left < right && arr[left] < base) {
      left++;
    }
    if (left < right) {
      arr[right--] = arr[left];
    }
  }
  arr[left] = base;
  return left;
}

function start(left, right) {
  if (left > right) {
    return;
  }
  const base = quickSort(left, right);
  start(left, base - 1);
  start(base + 1, right);
}

start(0, arr.length - 1);
console.log(arr);
